### PR TITLE
documentation context, lpae, vmm, mm, smp source code file

### PR DESCRIPTION
**Pull request by xiane**

### CODE REVIEW

**Comment by kesl** [comment]
Can one of the admins verify this patch?

**Comment by joohyun0115** [comment]
ok to test

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/77/

**Comment by in1004kyu** [comment]
First letter be capitalized, please.

File: hypervisor/context.c [hunk 1]
```diff
@@ -36,8 +39,12 @@
 #define CPSR_MODE_HYP   0x1A
 #define CPSR_MODE_UND   0x1B
 #define CPSR_MODE_SYS   0x1F
+/** @} */
 
 #define __CONTEXT_TRACE_VERBOSE__
+/**
+ * @brief check validattion of vmid
```
Comment (in1004kyu) [inline hypervisor/context.c]: -> Check validation of vmid.
First letter be capitalized, please.

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/78/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/79/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/81/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/82/

File: hypervisor/context.c [hunk 2]
```diff
@@ -109,6 +121,13 @@ void context_dump_regs(struct arch_regs *regs)
 #endif
 #endif
 }
+/**
+ * @brief Copy the register from regs_src to regs_dst
+ *
+ * copy registers cpsr, pc, lr, gpr[0-ARCH_REGS_NUM_GPR]
```
Comment (Wonseok) [inline hypervisor/context.c]: s/gpr[0-ARCH_REGS_NUM_GPR]/r0-r12

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/83/

File: hypervisor/context.c [hunk 3]
```diff
@@ -147,6 +172,13 @@ void context_init_banked(struct arch_regs_banked *regs_banked)
     /* Cortex-A15 processor does not support sp_fiq */
 }
 
+/**
+ * @brief Copy currently active banked registers to regs_banked
```
Comment (Wonseok) [inline hypervisor/context.c:176]: What does "currently active" mean?

File: hypervisor/context.c [hunk 4]
```diff
@@ -247,7 +285,11 @@ void context_restore_banked(struct arch_regs_banked *regs_banked)
                  : : "r"(regs_banked->r12_fiq) : "memory", "cc");
 }
 
-/* Co-processor state management: init/save/restore */
+/**
+ * @brief Co-processor state management: init
```
Comment (Wonseok) [inline hypervisor/context.c]: Please use Coprocessor instead of Co-processor.

Comment (Wonseok) [inline hypervisor/context.c]: What is the state management?

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/84/

File: hypervisor/context.c [hunk 5]
```diff
@@ -197,6 +229,12 @@ void context_save_banked(struct arch_regs_banked *regs_banked)
                  : "=r"(regs_banked->r12_fiq) : : "memory", "cc");
 }
 
+/**
+ * @brief Restore banked register from regs_banked to active banked register
```
Comment (Wonseok) [inline hypervisor/context.c:233]: Please do not use variable name. You might want to use general term, here.

File: hypervisor/context.c [hunk 6]
```diff
@@ -280,6 +332,19 @@ void context_restore_cops(struct arch_regs_cop *regs_cop)
    void context_switch_to_next_guest(struct arch_regs *regs_current)
  */
 
+/**
+ * @brief Switch context current to next guest context
```
Comment (Wonseok) [inline hypervisor/context.c:336]: Wrong description.

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/85/

File: hypervisor/context.c [hunk 7]
```diff
@@ -280,6 +332,19 @@ void context_restore_cops(struct arch_regs_cop *regs_cop)
...
+ *
+ * - Step<br>
+ *   - Disable Stage 2 Translation: HCR.VM = 0<br>
```
Comment (Wonseok) [inline hypervisor/context.c:339]: Please use uppercase letter in front of line only, except with acronym.

File: hypervisor/context.c [hunk 8]
```diff
@@ -280,6 +332,19 @@ void context_restore_cops(struct arch_regs_cop *regs_cop)
...
+ *   - Save the current Guest's context<br>
+ *   - Restore Translation Table for the next guest<br>
+ *   - and Enable Stage 2 Translation<br>
```
Comment (Wonseok) [inline hypervisor/context.c]: Remove and.

File: hypervisor/context.c [hunk 9]
```diff
@@ -280,6 +332,19 @@ void context_restore_cops(struct arch_regs_cop *regs_cop)
    void context_switch_to_next_guest(struct arch_regs *regs_current)
  */
 
+/**
+ * @brief Switch context current to next guest context
+ *
+ * - Step<br>
+ *   - Disable Stage 2 Translation: HCR.VM = 0<br>
+ *   - Save the current Guest's context<br>
+ *   - Restore Translation Table for the next guest<br>
```
Comment (Wonseok) [inline hypervisor/context.c:341]: We just only restore ttbr register, isn't it?

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/86/

File: hypervisor/context.h [hunk 10]
```diff
@@ -7,66 +7,117 @@
 #include "lpae.h"
 
 #define ARCH_REGS_NUM_GPR    13
-
+/**
+ * @brief Hypervisor hvc result
```
Comment (Wonseok) [inline hypervisor/context.h:11]: This is not only for hvc instruction. You might want to rename.

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/87/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/88/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/89/

File: hypervisor/context.h [hunk 11]
```diff
@@ -7,66 +7,117 @@
...
-    HYP_RESULT_ERET = 0,
-    HYP_RESULT_STAY = 1
+    HYP_RESULT_ERET = 0, /**< error status return */
+    HYP_RESULT_STAY = 1  /**< stay status return */
 };
 
+/**
+ * @brief Architecture Registers
+ *
+ * - Used to save the Current State Registers<br>
```
Comment (Wonseok) [inline hypervisor/context.h:21]: Wrong description.

File: hypervisor/context.h [hunk 12]
```diff
@@ -7,66 +7,117 @@
...
-    uint32_t gpr[ARCH_REGS_NUM_GPR]; /* R0 - R12 */
+    uint32_t cpsr;  /**< CPSR */
+    uint32_t pc;    /**< Program Counter */
+    uint32_t lr;    /**< Link Register */
+    uint32_t gpr[ARCH_REGS_NUM_GPR];  /**< R0 - R12 */
 } __attribute((packed));
 
-/* co-processor registers: cp15, cp2 */
+/**
+ *  @brief Co-Processor Registers: cp15, cp2
```
Comment (Wonseok) [inline hypervisor/context.h:35]: What is the cp2?

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/90/

File: hypervisor/context.h [hunk 13]
```diff
@@ -7,66 +7,117 @@
...
+    union lpaed *ttbl;    /**< Guest's translation table */
+    vmid_t vmid;    /**< Guest's Virtual machine ID */
+    struct vgic_status vgic_status;  /**< Guest's Virtual GIC Status */
+    struct arch_regs_cop regs_cop;  /**< Guest's Co-Processor Registers */
+    struct arch_regs_banked regs_banked;  /**< Guet's Banked Register */
 };
 
+/**
+ * @fn void context_dump_regs( struct arch_regs *regs)
+ * @brief Show delivered register's status
```
Comment (Wonseok) [inline hypervisor/context.h:90]: Wrong description.

File: hypervisor/context.h [hunk 14]
```diff
@@ -81,24 +132,95 @@ struct hyp_guest_context {
...
+ *   to switch the context,<br>
+ * where virq flush takes place, this time
+ * @param void
+ * @return hvmm_status_t result of swithcing
+ */
 hvmm_status_t context_perform_switch(void);
 
-/* VMID of the current guest */
+/**
+ * @brief VMID of the current guest
```
Comment (Wonseok) [inline hypervisor/context.h:152]: Wrong description.

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/91/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/92/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/93/

**Comment by kesl** [comment]
CI : Test FAILed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/94/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/95/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/96/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/97/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/98/

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/99/

**Pull request closed without merging**